### PR TITLE
Notification/feature/users

### PR DESCRIPTION
**Pull request by Shehab-m**
Users in notification service.

### CODE REVIEW

**Review by devsadeq — APPROVED** [review]

File: service_notification/src/main/kotlin/org/thechance/service_notification/data/collection/UserCollection.kt [hunk 1]
```diff
@@ -15,5 +15,8 @@ data class UserCollection(
     @SerialName("device_tokens")
     val deviceTokens: List<String>,
     @SerialName("topics")
-    val topics: List<Int>
+    val topics: List<Int>,
+    @SerialName("notifications")
+    val notifications: List<String>
```
Comment (mosayed01) [inline service_notification/src/main/kotlin/org/thechance/service_notification/data/collection/UserCollection.kt:20]: Is it notification IDs or notification titles ?
@Shehab-m

File: service_notification/src/main/kotlin/org/thechance/service_notification/domain/usecases/UserUseCase.kt [hunk 2]
```diff
@@ -0,0 +1,20 @@
+package org.thechance.service_notification.domain.usecases
+
+import org.koin.core.annotation.Single
+import org.thechance.service_notification.domain.gateway.UserGateway
+import org.thechance.service_notification.domain.model.User
+
+@Single
+class UserUseCase(private val userGateway: UserGateway): IUserUseCase {
```
Comment (mosayed01) [inline service_notification/src/main/kotlin/org/thechance/service_notification/domain/usecases/UserUseCase.kt:8]: What about user management use case?

File: service_notification/src/main/kotlin/org/thechance/service_notification/plugins/Routing.kt [hunk 3]
```diff
@@ -1,12 +1,14 @@
...
-import io.ktor.server.application.*
+import io.ktor.server.application.Application
+import io.ktor.server.routing.routing
 import org.thechance.service_notification.endpoints.testRoutes
+import org.thechance.service_notification.endpoints.userRoutes
 
 fun Application.configureRouting(
 ) {
     routing {
         testRoutes()
```
Comment (mosayed01) [inline service_notification/src/main/kotlin/org/thechance/service_notification/plugins/Routing.kt:11]: No need for test rout any more

**Review by mosayed01 — CHANGES_REQUESTED** [review]

File: service_notification/src/main/kotlin/org/thechance/service_notification/endpoints/model/UserDto.kt [hunk 4]
```diff
@@ -0,0 +1,11 @@
+package org.thechance.service_notification.endpoints.model
+
+import kotlinx.serialization.Serializable
+
+@Serializable
+data class UserDto(
+    val id: String,
+    val deviceTokens: List<String>,
+    val topics: List<Int>,
+    val notifications: List<String>
```
Comment (NadaFeteiha) [inline service_notification/src/main/kotlin/org/thechance/service_notification/endpoints/model/UserDto.kt:10]: I'm not sure what you mean by "notification". Do you mean the ID, the content, or the title? I also don't know what you're supposed to save or what this list represents!

File: service_notification/src/main/kotlin/org/thechance/service_notification/domain/usecases/IUserUseCase.kt [hunk 5]
```diff
@@ -0,0 +1,9 @@
+package org.thechance.service_notification.domain.usecases
+
+import org.thechance.service_notification.domain.model.User
+
+interface IUserUseCase {
+    suspend fun createUser(user: User): Boolean
```
Comment (NadaFeteiha) [inline service_notification/src/main/kotlin/org/thechance/service_notification/domain/usecases/IUserUseCase.kt:6]: I believe Bareq said that interfaces should be in the same file as their implementation.

File: service_notification/src/main/kotlin/org/thechance/service_notification/data/collection/UserCollection.kt (hunk 1 above)
Comment (Shehab-m) [reply service_notification/src/main/kotlin/org/thechance/service_notification/data/collection/UserCollection.kt:20]: ids @mosayed01

File: service_notification/src/main/kotlin/org/thechance/service_notification/plugins/Routing.kt (hunk 3 above)
Comment (Shehab-m) [reply service_notification/src/main/kotlin/org/thechance/service_notification/plugins/Routing.kt:11]: done

File: service_notification/src/main/kotlin/org/thechance/service_notification/domain/usecases/UserUseCase.kt (hunk 2 above)
Comment (Shehab-m) [reply service_notification/src/main/kotlin/org/thechance/service_notification/domain/usecases/UserUseCase.kt:8]: what's its use? @mosayed01

File: service_notification/src/main/kotlin/org/thechance/service_notification/endpoints/model/UserDto.kt (hunk 4 above)
Comment (Shehab-m) [reply service_notification/src/main/kotlin/org/thechance/service_notification/endpoints/model/UserDto.kt:10]: it's supposed to be ids got it changed.

Comment (Shehab-m) [reply service_notification/src/main/kotlin/org/thechance/service_notification/endpoints/model/UserDto.kt:10]: do you mean the naming of notifications should be changed to more explicit one to represent it's notifications ids ? @NadaFeteiha

Comment (NadaFeteiha) [reply service_notification/src/main/kotlin/org/thechance/service_notification/endpoints/model/UserDto.kt:10]: @Shehab-m Please ignore my previous comment, as it is the same as @mosayed01 's. I did not see his comment while I was reviewing the code.

Comment (Shehab-m) [reply service_notification/src/main/kotlin/org/thechance/service_notification/endpoints/model/UserDto.kt:10]: no worries thanks for your review.

File: service_notification/src/main/kotlin/org/thechance/service_notification/domain/usecases/UserUseCase.kt (hunk 2 above)
Comment (mosayed01) [reply service_notification/src/main/kotlin/org/thechance/service_notification/domain/usecases/UserUseCase.kt]: It's story to manage users notifications

File: service_notification/src/main/kotlin/org/thechance/service_notification/endpoints/userRoutes.kt [hunk 6]
```diff
@@ -23,12 +24,18 @@ fun Route.userRoutes() {
             val result = useCase.createUser(user.toEntity())
             call.respond(HttpStatusCode.Created, "User was created: $result")
         }
+        post("/{userId}/token") {
+            val userId = call.parameters["userId"] ?: throw BadRequestException("User id is required")
```
Comment (Asiasama710) [inline service_notification/src/main/kotlin/org/thechance/service_notification/endpoints/userRoutes.kt:28]: Can you improve the name of exception,  Hmmm  to MissingParameterException for example,i think is better than BadRequestException
oh and i see is better that using error code that be Int not string like this "User id is required"

Comment (Shehab-m) [reply service_notification/src/main/kotlin/org/thechance/service_notification/endpoints/userRoutes.kt:28]: Still gonna handle the validatin and errors.

**Review by mosayed01 — APPROVED** [review]
Well done

**Pull request merged**